src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -34456,6 +34440,455 @@ public class Derived : Base
             comp.VerifyDiagnostics();
         }
 
+        [Fact]
+        public void UnannotatedParam_UpdatesArgumentState()
+        {
+            var source = @"
+public class Program
+{
+    void M0(string s0) { }
+
+    void M1(string? s1)
+    {
+        M0(s1); // 1
+        _ = s1.ToString();
+    }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,12): warning CS8604: Possible null reference argument for parameter 's0' in 'void Program.M0(string s0)'.
+                //         M0(s1); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "s1").WithArguments("s0", "void Program.M0(string s0)").WithLocation(8, 12)
+                );
+        }
+
+        [Fact]
+        public void UnannotatedTypeArgument_UpdatesArgumentState()
+        {
+            var source = @"
+public class Program
+{
+    void M0<T>(T t) { }
+
+    void M1<T>(T? t) where T : class
+    {
+        M0<T>(t); // 1
+        _ = t.ToString();
+    }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,15): warning CS8604: Possible null reference argument for parameter 't' in 'void Program.M0<T>(T t)'.
+                //         M0<T>(t); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "t").WithArguments("t", "void Program.M0<T>(T t)").WithLocation(8, 15)
+                );
+        }
+
+        [Fact]
+        public void UnannotatedTypeArgument_NullableClassConstrained_DoesNotUpdateArgumentState()
```

Comment:
Consider making this test a `[Theory]` and using an argument to enumerate all the different constraint combinations where this is now valid.